### PR TITLE
Use span-based overloads

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

- First commit - Use span-based overloads to reduce allocations.
- Second commit - Use String.Contains() for readability.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/ModuleViewModel.cs [hunk 1]
```diff
@@ -442,14 +442,12 @@ private static bool MatchesEvenIfInPlural(string commandName, string filter)
```
Comment (PoshChan) [inline src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/ModuleViewModel.cs]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-macos`
Test-Connection.TraceRoute.TraceRoute works
```powershell
Expected exactly '10.79.1.108', but got 127.0.0.1.
at <ScriptBlock>, /Users/runner/runners/2.164.8/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Management/Test-Connection.Tests.ps1: line 257
257:             $result[0].TargetAddress | Should -BeExactly $realAddress
```

File: src/System.Management.Automation/engine/ParameterBinderController.cs [hunk 2]
```diff
@@ -345,7 +345,7 @@ internal static void AddArgumentsToCommandProcessor(CommandProcessorBase command
                             {
                                 param = CommandParameterInternal.CreateParameterWithArgument(
                                     /*parameterAst*/null, paramText.Substring(1, colonIndex - 1), paramText,
-                                    /*argumentAst*/null, paramText.Substring(colonIndex + 1).Trim(),
+                                    /*argumentAst*/null, paramText.AsSpan(colonIndex + 1).Trim().ToString(),
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/ParameterBinderController.cs]: ToString() here results in an allocation, can we just update that method to accept a readonlyspan?

Comment (iSazonov) [reply src/System.Management.Automation/engine/ParameterBinderController.cs]: No, it is used then in ctor-s.

**Comment by iSazonov** [comment]
Added new commit.

File: src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/CommandViewModel.cs [hunk 3]
```diff
@@ -429,7 +429,7 @@ public string GetScript()
                 commandName = this.ModuleName + "\\" + commandName;
```
Comment (PoshChan) [inline src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/CommandViewModel.cs]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-linux`
Default enablement of Experimental Features.On preview builds, Experimental Features are enabled
```powershell
Expected: Feature PSCommandNotFoundSuggestion to be Enabled
at <ScriptBlock>, /home/vsts/work/1/s/test/powershell/engine/ExperimentalFeature/Get-ExperimentalFeature.Tests.ps1: line 177
177:             $expFeature.Enabled | Should -BeEnabled -Name $expFeature.Name
```

Comment (PoshChan) [inline src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/CommandViewModel.cs]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-macos`
Default enablement of Experimental Features.On preview builds, Experimental Features are enabled
```powershell
Expected: Feature PSCommandNotFoundSuggestion to be Enabled
at <ScriptBlock>, /Users/runner/runners/2.164.8/work/1/s/test/powershell/engine/ExperimentalFeature/Get-ExperimentalFeature.Tests.ps1: line 177
177:             $expFeature.Enabled | Should -BeEnabled -Name $expFeature.Name
```

Comment (PoshChan) [inline src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/CommandViewModel.cs]: @iSazonov, your last commit had 1 failures in `PowerShell-CI-windows`
Default enablement of Experimental Features.On preview builds, Experimental Features are enabled
```powershell
Expected: Feature PSCommandNotFoundSuggestion to be Enabled
at <ScriptBlock>, D:\a\1\s\test\powershell\engine\ExperimentalFeature\Get-ExperimentalFeature.Tests.ps1: line 177
177:             $expFeature.Enabled | Should -BeEnabled -Name $expFeature.Name
```

File: src/System.Management.Automation/engine/ParameterBinderController.cs (hunk 2 above)
Comment (vexx32) [reply src/System.Management.Automation/engine/ParameterBinderController.cs]: `ReadOnlySpan<char>` typically is an implicit cast from `string`, isn't it?

Comment (iSazonov) [reply src/System.Management.Automation/engine/ParameterBinderController.cs]: @vexx32 No, only string -> ReadOnlySpan<char>

Comment (vexx32) [reply src/System.Management.Automation/engine/ParameterBinderController.cs]: Right, so if the method parameter is currently using `string` we should be able to change it to ReadOnlySpan<char> and have the implicit cast handle it, shouldn't we? Or am I missing some context here? 🙂

Comment (iSazonov) [reply src/System.Management.Automation/engine/ParameterBinderController.cs]: >  I missing some context here

CreateParameterWithArgument() passes the string to a constructor so it makes no sense to convert to span here.

**Review by SteveL-MSFT — APPROVED** [review]
LGTM

File: src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/ModuleViewModel.cs [hunk 4]
```diff
@@ -435,21 +435,19 @@ private static bool Matches(WildcardPattern filterPattern, string commandName, s
...
+            if (commandName.Contains(filter, StringComparison.OrdinalIgnoreCase))
             {
                 return true;
             }
 
             if (filter.Length > 5 && filter.EndsWith("es", StringComparison.OrdinalIgnoreCase))
             {
-                filter = filter.Substring(0, filter.Length - 2);
-                return commandName.IndexOf(filter, StringComparison.OrdinalIgnoreCase) != -1;
+                return commandName.AsSpan().Contains(filter.AsSpan(0, filter.Length - 2), StringComparison.OrdinalIgnoreCase);
```
Comment (daxian-dbw) [inline src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/ModuleViewModel.cs]: Please use multiple statements for better readability, like
```
var filterSpan = filter.AsSpan(0, filter.Length - 2);
return commandName.AsSpan().Contains(filterSpan, StringComparison.OrdinalIgnoreCase);
```

File: src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/ModuleViewModel.cs [hunk 5]
```diff
@@ -435,21 +435,19 @@ private static bool Matches(WildcardPattern filterPattern, string commandName, s
...
             }
 
             if (filter.Length > 4 && filter.EndsWith("s", StringComparison.OrdinalIgnoreCase))
             {
-                filter = filter.Substring(0, filter.Length - 1);
-                return commandName.IndexOf(filter, StringComparison.OrdinalIgnoreCase) != -1;
+                return commandName.AsSpan().Contains(filter.AsSpan(0, filter.Length - 1), StringComparison.OrdinalIgnoreCase);
```
Comment (daxian-dbw) [inline src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/ModuleViewModel.cs]: same here :)

File: src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/ModuleViewModel.cs (hunk 4 above)
Comment (iSazonov) [reply src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/ModuleViewModel.cs]: Fixed.

File: src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/ModuleViewModel.cs (hunk 5 above)
Comment (iSazonov) [reply src/Microsoft.Management.UI.Internal/ShowCommand/ViewModel/ModuleViewModel.cs]: Fixed.

**Pull request merged**